File: netlify/functions/stripe-webhook.js
/**
 * stripe-webhook.js
 * POST /.netlify/functions/stripe-webhook
 *
 * Receives and verifies Stripe webhook events.
 *
 * Events handled:
 *   checkout.session.completed     → create client in Supabase + trigger onboarding
 *   customer.subscription.deleted  → mark client cancelled in Supabase
 *
 * Register this endpoint in Stripe Dashboard → Developers → Webhooks:
 *   https://getsprintai.com/.netlify/functions/stripe-webhook
 */

const stripe = require("stripe")(process.env.STRIPE_SECRET_KEY);
const { createClient } = require("@supabase/supabase-js");

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY
);

const PLAN_NAMES = {
  [process.env.STRIPE_PRICE_FOUNDER]: "founder",
  [process.env.STRIPE_PRICE_GROWTH]: "growth",
};

// ─── Helpers ──────────────────────────────────────────────────────────────────

async function upsertClient({ email, name, plan, stripeCustomerId }) {
  const { data: existing } = await supabase
    .from("sprintai_clients")
    .select("id")
    .eq("email", email)
    .single();

  if (existing) {
    await supabase
      .from("sprintai_clients")
      .update({ name, plan, stripe_customer_id: stripeCustomerId, status: "active" })
      .eq("id", existing.id);
    console.log(`Updated client: ${email} (${existing.id})`);
    return existing.id;
  }

  const { data } = await supabase
    .from("sprintai_clients")
    .insert({ email, name, plan, stripe_customer_id: stripeCustomerId, status: "active" })
    .select("id")
    .single();
  console.log(`Created client: ${email} (${data.id})`);
  return data.id;
}

async function getPlanFromSession(session) {
  const meta = session.metadata || {};
  if (meta.plan) return meta.plan;

  const subId = session.subscription;
  if (subId) {
    try {
      const sub = await stripe.subscriptions.retrieve(subId, {
        expand: ["items.data.price"],
      });
      const priceId = sub.items.data[0].price.id;
      return PLAN_NAMES[priceId] || priceId;
    } catch (e) {
      console.warn("Could not retrieve subscription plan:", e.message);
    }
  }
  return "unknown";
}

// ─── Handler ──────────────────────────────────────────────────────────────────

exports.handler = async (event) => {
  if (event.httpMethod !== "POST") {
    return { statusCode: 405, body: "Method not allowed" };
  }

  // Stripe needs the raw body for signature verification
  const rawBody = event.isBase64Encoded
    ? Buffer.from(event.body, "base64")
    : event.body;

  const sig = event.headers["stripe-signature"];

  let stripeEvent;
  try {
    stripeEvent = stripe.webhooks.constructEvent(
      rawBody,
      sig,
      process.env.STRIPE_WEBHOOK_SECRET
    );
  } catch (err) {
    console.error("Webhook signature verification failed:", err.message);
    return { statusCode: 400, body: `Webhook Error: ${err.message}` };
  }

  console.log(`Stripe event: ${stripeEvent.type} (${stripeEvent.id})`);

  try {
    // ── checkout.session.completed ────────────────────────────────────────────
    if (stripeEvent.type === "checkout.session.completed") {
      const session = stripeEvent.data.object;
      const details = session.customer_details || {};
      const email = session.customer_email || details.email || "";
      const name = details.name || "";
      const stripeCustomerId = session.customer || "";
      const plan = await getPlanFromSession(session);

      if (!email) {
        console.error("No email in checkout.session.completed — session:", session.id);
        return { statusCode: 200, body: "ok (no email)" };
      }

      const clientId = await upsertClient({ email, name, plan, stripeCustomerId });
      console.log(`Client ready: ${email} | plan: ${plan} | id: ${clientId}`);

      // TODO: trigger onboarding email once SMTP is configured
      // For now: client is in Supabase + redirected to /welcome by Stripe
    }

    // ── customer.subscription.deleted ─────────────────────────────────────────
    else if (stripeEvent.type === "customer.subscription.deleted") {
      const subscription = stripeEvent.data.object;
      const customerId = subscription.customer;

      const { data } = await supabase
        .from("sprintai_clients")
        .update({ status: "cancelled" })
        .eq("stripe_customer_id", customerId)
        .select("email");

      if (data && data.length > 0) {
        console.log(`Cancelled subscription for: ${data[0].email}`);
      } else {
        console.warn(`No client found for Stripe customer: ${customerId}`);
      }
    }
  } catch (err) {
    console.error("Error processing webhook:", err);
    // Return 200 anyway — Stripe will retry if we return non-2xx
  }

  return { statusCode: 200, body: "ok" };
};

File: projects/sprintai/backend/stripe_webhook.py
"""
stripe_webhook.py — SprintAI Stripe webhook handler + checkout session creator.

Routes
------
POST /create-checkout-session  →  Creates a Stripe Checkout session and returns the URL.
POST /webhook                  →  Receives Stripe webhook events.

Events handled
--------------
  checkout.session.completed      → create client record in Supabase + send onboarding email
  customer.subscription.deleted  → mark client status = 'cancelled' in Supabase

Run locally
-----------
  pip install flask stripe supabase python-dotenv
  export FLASK_APP=stripe_webhook.py
  flask run --port 4242

  # In another terminal, forward Stripe events:
  stripe listen --forward-to localhost:4242/webhook
"""

import os
import sys
import subprocess
import logging
from flask import Flask, request, jsonify
import stripe
from supabase import create_client, Client
from dotenv import load_dotenv

load_dotenv()

# ─── Logging ────────────────────────────────────────────────────────────────
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
)
logger = logging.getLogger(__name__)

# ─── Config ─────────────────────────────────────────────────────────────────
STRIPE_SECRET_KEY      = os.environ["STRIPE_SECRET_KEY"]
STRIPE_WEBHOOK_SECRET  = os.environ["STRIPE_WEBHOOK_SECRET"]
STRIPE_PRICE_FOUNDER   = os.environ.get("STRIPE_PRICE_FOUNDER", "PRICE_FOUNDER")
STRIPE_PRICE_GROWTH    = os.environ.get("STRIPE_PRICE_GROWTH",  "PRICE_GROWTH")

SUPABASE_URL           = os.environ["SUPABASE_URL"]
SUPABASE_SERVICE_KEY   = os.environ["SUPABASE_SERVICE_ROLE_KEY"]

SUCCESS_URL            = os.environ.get(
    "CHECKOUT_SUCCESS_URL",
    "https://getsprintai.com/welcome?session_id={CHECKOUT_SESSION_ID}"
)
CANCEL_URL             = os.environ.get(
    "CHECKOUT_CANCEL_URL",
    "https://getsprintai.com/checkout"
)

stripe.api_key = STRIPE_SECRET_KEY

supabase: Client = create_client(SUPABASE_URL, SUPABASE_SERVICE_KEY)

app = Flask(__name__)


# ─── Helpers ────────────────────────────────────────────────────────────────

PLAN_NAMES = {
    STRIPE_PRICE_FOUNDER: "founder",
    STRIPE_PRICE_GROWTH:  "growth",
}


def _plan_from_session(session: dict) -> str:
    """Derive plan name from session metadata or line-item price ID."""
    # Prefer explicit metadata set at session creation time
    meta = session.get("metadata") or {}
    if meta.get("plan"):
        return meta["plan"]

    # Fall back to the first subscription item's price
    subscription_id = session.get("subscription")
    if subscription_id:
        try:
            sub = stripe.Subscription.retrieve(
                subscription_id,
                expand=["items.data.price"],
            )
            price_id = sub["items"]["data"][0]["price"]["id"]
            return PLAN_NAMES.get(price_id, price_id)
        except stripe.error.StripeError as exc:
            logger.warning("Could not retrieve subscription plan: %s", exc)

    return meta.get("plan", "unknown")


def _upsert_client(
    email: str,
    name: str,
    plan: str,
    stripe_customer_id: str,
    status: str = "active",
) -> str:
    """Insert or update a client in sprintai_clients. Returns the client UUID."""
    existing = (
        supabase.table("sprintai_clients")
        .select("id")
        .eq("email", email)
        .execute()
    )
    if existing.data:
        client_id = existing.data[0]["id"]
        supabase.table("sprintai_clients").update(
            {
                "name": name,
                "plan": plan,
                "stripe_customer_id": stripe_customer_id,
                "status": status,
            }
        ).eq("id", client_id).execute()
        logger.info("Updated existing client %s (%s)", email, client_id)
    else:
        result = (
            supabase.table("sprintai_clients")
            .insert(
                {
                    "email": email,
                    "name": name,
                    "plan": plan,
                    "stripe_customer_id": stripe_customer_id,
                    "status": status,
                }
            )
            .execute()
        )
        client_id = result.data[0]["id"]
        logger.info("Created new client %s (%s)", email, client_id)

    return client_id


def _send_onboarding_email(email: str, name: str, client_id: str) -> None:
    """Fire the onboarding email script as a subprocess."""
    script = os.path.join(os.path.dirname(__file__), "send_onboarding_email.py")
    cmd = [
        sys.executable, script,
        "--client_email", email,
        "--client_name",  name,
        "--client_id",    client_id,
    ]
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        logger.error("Onboarding email failed: %s", result.stderr)
    else:
        logger.info("Onboarding email sent to %s", email)


# ─── Routes ─────────────────────────────────────────────────────────────────

@app.route("/create-checkout-session", methods=["POST"])
def create_checkout_session():
    """
    Create a Stripe Checkout session for the requested plan.

    Request JSON:
        { "plan": "founder" | "growth", "price_id": "<stripe_price_id>" }

    Response JSON:
        { "url": "<stripe_checkout_url>" }
    """
    data = request.get_json(silent=True) or {}
    plan     = data.get("plan", "unknown")
    price_id = data.get("price_id")

    if not price_id:
        return jsonify({"error": "price_id is required"}), 400

    try:
        session = stripe.checkout.Session.create(
            payment_method_types=["card"],
            mode="subscription",
            line_items=[{"price": price_id, "quantity": 1}],
            success_url=SUCCESS_URL,
            cancel_url=CANCEL_URL,
            metadata={"plan": plan},
            # Collect billing name + email so they appear in the webhook event
            billing_address_collection="auto",
            customer_creation="always",
        )
        return jsonify({"url": session.url})

    except stripe.error.StripeError as exc:
        logger.error("Stripe error creating session: %s", exc)
        return jsonify({"error": str(exc)}), 500


@app.route("/webhook", methods=["POST"])
def stripe_webhook():
    """
    Receive and verify Stripe webhook events.

    Verifies the Stripe-Signature header against STRIPE_WEBHOOK_SECRET,
    then dispatches to the appropriate handler.
    """
    payload   = request.get_data()
    sig_header = request.headers.get("Stripe-Signature", "")

    try:
        event = stripe.Webhook.construct_event(
            payload, sig_header, STRIPE_WEBHOOK_SECRET
        )
    except ValueError:
        logger.warning("Invalid payload from Stripe")
        return jsonify({"error": "Invalid payload"}), 400
    except stripe.error.SignatureVerificationError:
        logger.warning("Invalid Stripe signature")
        return jsonify({"error": "Invalid signature"}), 400

    event_type = event["type"]
    logger.info("Received Stripe event: %s  id=%s", event_type, event["id"])

    # ── checkout.session.completed ─────────────────────────────────────────
    if event_type == "checkout.session.completed":
        session = event["data"]["object"]

        email    = session.get("customer_email") or session.get("customer_details", {}).get("email", "")
        name     = (session.get("customer_details") or {}).get("name", "")
        customer = session.get("customer", "")
        plan     = _plan_from_session(session)

        if not email:
            logger.error("No email in checkout.session.completed — session id=%s", session["id"])
            return jsonify({"error": "Missing customer email"}), 200  # return 200 to ack Stripe

        try:
            client_id = _upsert_client(email, name, plan, customer)
            _send_onboarding_email(email, name or "there", client_id)
        except Exception as exc:
            logger.exception("Error handling checkout.session.completed: %s", exc)
            # Still return 200 so Stripe doesn't retry indefinitely.
            # Alert should be raised via logging/monitoring.

    # ── customer.subscription.deleted ─────────────────────────────────────
    elif event_type == "customer.subscription.deleted":
        subscription = event["data"]["object"]
        customer_id  = subscription.get("customer", "")

        try:
            result = (
                supabase.table("sprintai_clients")
                .update({"status": "cancelled"})
                .eq("stripe_customer_id", customer_id)
                .execute()
            )
            if result.data:
                logger.info("Cancelled subscription for customer %s", customer_id)
            else:
                logger.warning("No client found for customer %s (subscription deleted)", customer_id)
        except Exception as exc:
            logger.exception("Error handling customer.subscription.deleted: %s", exc)

    else:
        logger.debug("Unhandled event type: %s", event_type)

    return jsonify({"status": "ok"}), 200


# ─── Entry point ────────────────────────────────────────────────────────────

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 4242))
    debug = os.environ.get("FLASK_DEBUG", "false").lower() == "true"
    logger.info("Starting SprintAI webhook server on port %d", port)
    app.run(host="0.0.0.0", port=port, debug=debug)
